### PR TITLE
Improve auto-evo performance by caching

### PR DESCRIPTION
**Pull request by Maxonovien**
**Brief Description of What This PR Does**

This PR adds further caching for auto-evo computations. Last checks indicated a reduction from 4m 22s to 48s to compute 20 generations.

**Related Issues**

None, but improving auto-evo performance is always good.

**Progress Checklist**

Note: before starting this checklist the PR should be marked as non-draft.

- [x] PR author has checked that this PR works as intended and doesn't
      break existing features:
      https://wiki.revolutionarygamesstudio.com/wiki/Testing_Checklist
      (this is important as to not waste the time of Thrive team
      members reviewing this PR)
- [x] Initial code review passed (this and further items should not be checked by the PR author)
- [x] Functionality is confirmed working by another person (see above checklist link)
- [x] Final code review is passed and code conforms to the 
      [styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md).

Before merging all CI jobs should finish on this PR without errors, if
there are automatically detected style issues they should be fixed by
the PR author. Merging must follow our
[styleguide](https://github.com/Revolutionary-Games/Thrive/blob/master/doc/style_guide.md#git).

### CODE REVIEW

File: src/auto-evo/simulation/SimulationCache.cs [hunk 1]
```diff
@@ -23,6 +29,12 @@ public class SimulationCache
         private readonly Dictionary<(TweakedProcess, BiomeConditions), ProcessSpeedInformation> cachedProcessSpeeds =
             new();
 
+        private readonly Dictionary<MicrobeSpecies, float[]> cachedPredationToolsRawScores = new();
+        private readonly Dictionary<(OrganelleTemplate, BiomeConditions, Compound), float>
```
Comment (hhyyrylainen) [inline src/auto-evo/simulation/SimulationCache.cs:33]: I think this cache size will blow up pretty easily as each species has one OrganelleTemplate per organelle, and those instances are not equal when new attempted organelles are used. Would caching based on OrganelleDefinition work? Those wouldn't pollute the cache with single use data.

File: src/auto-evo/simulation/food_source/HeterotrophicFoodSource.cs [hunk 2]
```diff
@@ -68,30 +68,12 @@ public HeterotrophicFoodSource(Patch patch, MicrobeSpecies prey, SimulationCache
...
-                    }
-
-                    if (process.Process.Outputs.TryGetValue(mucilage, out var mucilageAmount))
-                    {
-                        mucilageScore += mucilageAmount * Constants.AUTO_EVO_MUCILAGE_PREDATION_SCORE;
-                    }
-                }
-            }
+            // The array was used to have a time & memory efficient structure,
+            // but this could be turned into a class for more readable and less error-prone code.
```
Comment (hhyyrylainen) [inline src/auto-evo/simulation/food_source/HeterotrophicFoodSource.cs:72]: What about a tuple with named elements? Those are used a lot in Thrive to quickly package a couple of variables into one unit.

File: src/auto-evo/simulation/SimulationCache.cs (hunk 1 above)
Comment (Maxonovien) [reply src/auto-evo/simulation/SimulationCache.cs:33]: I think it would. I'll need to have a closer look at it to be sure though.

**Comment by buckly90** [comment]
With this change, most players may not ever have to wait on autoevo during normal gameplay without the help of cheats. Wonderful work!

**Review by hhyyrylainen — APPROVED** [review]
Code change looks good. I think the cache size is now not going to blow up. I tested this and I got just about 5% higher memory usage when just using the auto-evo exploring tool. I ran 20 generations for 10 worlds (I got around 670 MB memory usage). And for the performance I got about exactly 50% of the previous auto-evo runtime so I'd say that's a pretty huge improvement (from around 4:48 to 2:27).

**Pull request merged**